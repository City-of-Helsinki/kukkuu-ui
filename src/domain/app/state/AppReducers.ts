import { combineReducers } from '@reduxjs/toolkit';

import authenticationReducers from '../../auth/state/AuthenticationReducers';
import profileReducer from '../../profile/state/ProfileReducers';
import registrationReducer from '../../registration/state/RegistrationReducers';
import uiReducers from './ui/UIReducers';
import EventReducers from '../../event/state/EventReducers';

export default combineReducers({
  authentication: authenticationReducers,
  event: EventReducers,
  registration: registrationReducer,
  profile: profileReducer,
  ui: uiReducers,
});
