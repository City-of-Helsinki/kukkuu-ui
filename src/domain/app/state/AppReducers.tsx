import { combineReducers } from 'redux-starter-kit';
import { reducer as oidcReducer } from 'redux-oidc';

import ApiReducer from '../../auth/state/ApiReducer';
import registrationReducer from '../../registration/state/RegistrationReducers';

const authenticationReducer = combineReducers({
  apiAuthentication: ApiReducer,
  oidcReducer,
});

const rootReducer = combineReducers({
  authentication: authenticationReducer,
  registration: registrationReducer,
});

export type RootState = ReturnType<typeof rootReducer>;
export default rootReducer;
