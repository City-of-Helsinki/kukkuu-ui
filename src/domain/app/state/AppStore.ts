import { configureStore, getDefaultMiddleware, Store } from 'redux-starter-kit';
import { persistStore, persistReducer } from 'redux-persist';
import storage from 'redux-persist/lib/storage';

import rootReducer from './AppReducers';

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['registration'],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store: Store = configureStore({
  devTools: true,
  middleware: getDefaultMiddleware({
    serializableCheck: false,
  }),
  reducer: persistedReducer,
});

const persistor = persistStore(store);

export { persistor, store };
