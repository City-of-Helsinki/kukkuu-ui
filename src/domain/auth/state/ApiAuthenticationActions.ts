import { createAction } from 'redux-starter-kit';

import { API_AUTHENTICATION_ACTIONS } from '../constants/ApiAuthenticationActionConstants';

const startFetchingToken = createAction(
  API_AUTHENTICATION_ACTIONS.START_FETCHING_TOKEN
);

const fetchTokenSuccess = createAction(
  API_AUTHENTICATION_ACTIONS.FETCH_TOKEN_SUCCESS
);

const fetchTokenError = createAction(
  API_AUTHENTICATION_ACTIONS.FETCH_TOKEN_ERROR
);

export { startFetchingToken, fetchTokenSuccess, fetchTokenError };
