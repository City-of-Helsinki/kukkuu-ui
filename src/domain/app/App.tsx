import { Route, Switch, Redirect, useParams } from 'react-router';
import React, { useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import 'react-toastify/dist/ReactToastify.css';

import Home from '../home/Home';
import NotFound from './notFound/NotFound';
import NotEligible from '../registration/notEligible/NotEligible';
import PrivateRoute from '../auth/route/PrivateRoute';
import RegistrationForm from '../registration/form/RegistrationForm';
import LoadingSpinner from '../../common/components/spinner/LoadingSpinner';
import {
  isLoadingUserSelector,
  apiTokenSelector,
  userSelector,
} from '../auth/state/AuthenticationSelectors';
import Welcome from '../registration/welcome/Welcome';
import Profile from '../profile/Profile';
import AccessibilityStatement from '../accessibilityStatement/AccessibilityStatement';
import { userHasProfileSelector } from '../registration/state/RegistrationSelectors';
import TermsOfService from '../termsOfService/TermsOfService';
import { authenticateWithBackend } from '../auth/authenticate';

const App: React.FunctionComponent = props => {
  const isLoadingUser = useSelector(isLoadingUserSelector);
  const { locale } = useParams<{ locale: string }>();
  const userHasProfile = useSelector(userHasProfileSelector);
  const apiToken = useSelector(apiTokenSelector);
  const dispatch = useDispatch();
  const user = useSelector(userSelector);

  useEffect(() => {
    if (!apiToken && user?.access_token) {
      dispatch(authenticateWithBackend(user.access_token));
    }
  }, [apiToken, dispatch, user]);

  return (
    <LoadingSpinner isLoading={isLoadingUser}>
      <Switch>
        <Redirect exact path={`/${locale}/`} to={`/${locale}/home`} />
        <Route exact path={`/${locale}/home`} component={Home} />
        <Route
          exact
          path={`/${locale}/registration/not-eligible`}
          component={NotEligible}
        />
        <Route
          exact
          path={`/${locale}/accessibility`}
          component={AccessibilityStatement}
        />
        <Route exact path={`/${locale}/terms`} component={TermsOfService} />
        {!userHasProfile && (
          <PrivateRoute exact path={`/${locale}/registration/form`}>
            <RegistrationForm />
          </PrivateRoute>
        )}
        <PrivateRoute exact path={`/${locale}/registration/success`}>
          <Welcome />
        </PrivateRoute>

        <PrivateRoute path={`/${locale}/profile`}>
          <Profile />
        </PrivateRoute>

        {userHasProfile && <Redirect to={`/${locale}/profile`} />}

        <Route component={NotFound} />
      </Switch>
    </LoadingSpinner>
  );
};

export default App;
