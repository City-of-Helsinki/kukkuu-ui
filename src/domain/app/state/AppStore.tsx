import {
  Action,
  configureStore,
  getDefaultMiddleware,
} from 'redux-starter-kit';
import { persistStore, persistReducer } from 'redux-persist';
import { ThunkAction } from 'redux-thunk';
import storage from 'redux-persist/lib/storage';
import { loadUser, USER_FOUND } from 'redux-oidc';

import userManager from '../../auth/userManager';
import { fetchApiTokenThunk } from '../../auth/state/ApiReducer';
import rootReducer, { RootState } from './AppReducers';

const isProd = process.env.NODE_ENV === 'production';

export type AppThunk = ThunkAction<void, RootState, null, Action<string>>;

const persistConfig = {
  key: 'root',
  storage,
  whitelist: ['registration'],
};

const persistedReducer = persistReducer(persistConfig, rootReducer);

const store = configureStore({
  devTools: !isProd,
  middleware: getDefaultMiddleware({
    serializableCheck: {
      ignoredActions: [USER_FOUND, 'persist/PERSIST'],
    },
  }),
  reducer: persistedReducer,
});

const persistor = persistStore(store);
loadUser(store, userManager).then(async user => {
  if (user) {
    store.dispatch(fetchApiTokenThunk(user.access_token));
  }
});

export { persistor, store };
