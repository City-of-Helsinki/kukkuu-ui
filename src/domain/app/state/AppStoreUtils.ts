import { persistor, store } from './AppStore';
import { resetBackendAuthentication } from '../../auth/state/BackendAuthenticationActions';
import client from '../../api/client';
import { clearProfile } from '../../profile/state/ProfileActions';
import { resetFormValues } from '../../registration/state/RegistrationActions';

/**
 * Flush entire Redux data store
 * as well as persisted data
 * Apollo cache
 */
export const flushData = () => {
  // Clear backend auth data
  store.dispatch(resetBackendAuthentication());
  // Clear user form data
  store.dispatch(resetFormValues());
  // Clear profile (fetched from API)
  store.dispatch(clearProfile());
  // Flush data in redux store and localStorage
  persistor.flush();
  // Clear Apollo cache
  client.clearStore();
};
