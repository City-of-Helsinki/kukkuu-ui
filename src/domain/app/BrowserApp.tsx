import React, { FunctionComponent } from 'react';
import { BrowserRouter } from 'react-router-dom';
import { Switch, Route, Redirect } from 'react-router';
import { Provider } from 'react-redux';
import { OidcProvider } from 'redux-oidc';

import App from './App';
import userManager from '../auth/userManager';
import enableOidcLogging from '../auth/enableOidcLogging';
import OidcCallback from '../auth/OidcCallback';
import { SUPPORT_LANGUAGES } from '../../common/translation/constants';
import store from './state/AppStore';

const localeParam = `:locale(${SUPPORT_LANGUAGES.EN}|${SUPPORT_LANGUAGES.FI}|${SUPPORT_LANGUAGES.SV})`;

if (process.env.NODE_ENV !== 'production') {
  enableOidcLogging();
}

// Export for testing purpose
export const appRoutes = (
  <Switch>
    <Route exact path="/callback" component={OidcCallback} />
    <Redirect exact path="/" to="/fi/home" />
    <Route path={`/${localeParam}/*`} component={App} />
    <Route exact path={`/${localeParam}/callback`} component={OidcCallback} />
    <Route
      render={props => <Redirect to={`/fi${props.location.pathname}`} />}
    />
  </Switch>
);
const BrowserApp: FunctionComponent = () => {
  return (
    <Provider store={store}>
      <OidcProvider store={store} userManager={userManager}>
        <BrowserRouter>{appRoutes}</BrowserRouter>
      </OidcProvider>
    </Provider>
  );
};

export default BrowserApp;
