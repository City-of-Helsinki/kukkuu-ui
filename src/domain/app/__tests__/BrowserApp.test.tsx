import React from 'react';
import { mount } from 'enzyme';
import { MemoryRouter } from 'react-router';
import { Provider } from 'react-redux';
import toJson from 'enzyme-to-json';
import { MockedProvider } from '@apollo/react-testing';
import { act } from 'react-dom/test-utils';

import { store } from '../state/AppStore';
import BrowserApp, { AppRoutes } from '../BrowserApp';
import App from '../App';

// Tell React not to wait for Apollo query to resolve
// eslint-disable-next-line @typescript-eslint/no-explicit-any
const updateWrapper = async (wrapper: any, time = 0) => {
  await act(async () => {
    await new Promise(res => setTimeout(res, time));
    await wrapper.update();
  });
};

const wrapperCreator = (route: string) => {
  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter initialEntries={[route]}>
        <MockedProvider>
          <AppRoutes />
        </MockedProvider>
      </MemoryRouter>
    </Provider>
  );
  updateWrapper(wrapper);

  return wrapper;
};

it('renders snapshot correctly', () => {
  const tree = mount(<BrowserApp />);
  expect(toJson(tree)).toMatchSnapshot();
});

it('redirect user from root to /fi/home by default', () => {
  const wrapper = wrapperCreator('/');
  expect(
    wrapper
      .children()
      .children()
      .props().history.location.pathname
  ).toBe('/fi/home');
});

it('user from root will be redirect to App with guarantee fi locale', () => {
  const wrapper = wrapperCreator('/');
  const app = wrapper.find(App);

  expect(app).toBeDefined();
  expect(app.props().match.params.locale).toEqual('fi');
});

it('user from supported locale will be redirect to App with that locale', () => {
  const wrapper = wrapperCreator('/en/');
  const app = wrapper.find(App);

  expect(app).toBeDefined();
  expect(app.props().match.params.locale).toEqual('en');
});

it('user from unsupported locale prefix will be redirect to route with support prefix', () => {
  const wrapper = wrapperCreator('/vi/');
  const app = wrapper.find(App);

  expect(app.props().match.params.locale).toEqual('fi');
  expect(app.props().location.pathname).toContain('/fi/vi/');
});

it('user without locale prefix will be redirect to route with support prefix', () => {
  const wrapper = wrapperCreator('/foo-url');
  const app = wrapper.find(App);

  expect(app.props().match.params.locale).toEqual('fi');
  expect(app.props().location.pathname).toContain('/fi/foo-url');
});

it('user with route with unsupport locale will be redirect to App anyway, with supported locale', () => {
  const wrapper = wrapperCreator('/dk/foo');
  const app = wrapper.find(App);

  expect(app.props().match.params.locale).toEqual('fi');
  expect(app.props().location.pathname).toContain('/fi/dk/foo');
});
