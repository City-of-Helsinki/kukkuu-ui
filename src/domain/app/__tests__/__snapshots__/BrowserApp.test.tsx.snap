// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders snapshot correctly 1`] = `"<div class=\\"spinnerWrapper\\"><div class=\\"pageWrapper\\"><header class=\\"headerWrapper\\"><div class=\\"container\\"><div class=\\"navbarTop\\"><div class=\\"logo\\"></div><h3 class=\\"appName\\">Kulttuurin Kummilapset</h3></div></div></header><div class=\\"pageBody\\"><div class=\\"container\\"><div class=\\"home\\"><div class=\\"hero\\"><div class=\\"heroContainer\\"><h1>Kulttuurin Kummilapset 2020</h1><p class=\\"description\\">Helsingin Kaupunginorkesteri kutsuu kummilapsikseen kaikki vuonna 2020 syntyneet Helsinkiläislapset! Lapsi perheineen kutsutaan kahdesti vuodessa nauttimaan KHO:n järjestämästä kulttuurista täysin ilmaiseksi! Liity mukaan!<span role=\\"button\\">Login using tunnistamo</span></p><div class=\\"homeForm\\"><form><div class=\\"inputWrapper\\"><div class=\\"birthdayField\\"><label>Lapsen syntymäaika</label><div class=\\"inputWrapper\\"><div class=\\"formInputField\\"><div class=\\"className inputWrapper numberInput\\"><div class=\\"inputTypeWrapper\\"><input type=\\"number\\" name=\\"childBirthdayDay\\" placeholder=\\"pv\\" min=\\"1\\" max=\\"31\\" value=\\"\\"></div><small class=\\"inputError\\"></small></div></div><div class=\\"dot\\">.</div><div class=\\"formInputField\\"><div class=\\"className inputWrapper numberInput\\"><div class=\\"inputTypeWrapper\\"><input type=\\"number\\" name=\\"childBirthdayMonth\\" placeholder=\\"kk\\" min=\\"1\\" max=\\"12\\" value=\\"\\"></div><small class=\\"inputError\\"></small></div></div><div class=\\"dot\\">.</div><div class=\\"formInputField\\"><div class=\\"className inputWrapper numberInput\\"><div class=\\"inputTypeWrapper\\"><input type=\\"number\\" name=\\"childBirthdayYear\\" placeholder=\\"vuosi\\" min=\\"2019\\" value=\\"\\"></div><small class=\\"inputError\\"></small></div></div></div><small class=\\"error\\"></small></div><div class=\\"formInputField\\"><div class=\\"className inputWrapper textInput\\"><div class=\\"inputTypeWrapper\\"><label>Lapsen kotipaikkakunta</label><input type=\\"text\\" name=\\"childHomeCity\\" placeholder=\\"Enter hometown city, currently only Helsinki supported\\" value=\\"\\"></div><small class=\\"inputError\\"></small></div></div></div><div class=\\"formInputField\\"><div class=\\"className inputWrapper checkboxInput\\"><div class=\\"inputTypeWrapper\\"><label>Vakuutan antamani tiedot oikeaksi.</label><input type=\\"checkbox\\" name=\\"verifyInformation\\" value=\\"false\\"></div><small class=\\"inputError\\"></small></div></div><div class=\\"submitButton buttonWrapper\\"><button type=\\"submit\\" disabled=\\"\\">Ilmoittaudu mukaan</button></div></form></div></div></div></div></div></div><footer class=\\"footerWrapper\\"><div class=\\"container\\"><div class=\\"footer\\"><div class=\\"helsinkiLogo\\"></div><div class=\\"copyright\\"><p>© Copyright 2019 • All right reserved •</p></div></div></div></footer></div></div>"`;
